### ISSUE TITLE
Webconsole fails to install on Opencast 12.x and previous

### ISSUE THREAD

**Issue by mliradelc** [opencast/opencast#4610]
**Describe the bug**

When you try to install the Karaf web console in any opencast instance that has the `opencast-engage-theodul-core` active, you get `java.lang.IllegalStateException: Service already unregistered.` error. Halting the installation.

**To Reproduce**

1. Build or install any Opencast version that has `opencast-engage-theodul-core` enabled (AllinOne for example).
2. In karaf shell, run `feature:install -v webconsole`
3. See error

**Expected behavior**

Webconsole should install without problems

**Stack trace**

```
ERROR : bundle opencast-engage-theodul-core:12.0.0.SNAPSHOT (108)[org.opencastproject.engage.theodul.manager.impl.EngagePluginManagerImpl(107)] : The deactivate method has thrown an exception
java.lang.IllegalStateException: Service already unregistered.
        at org.apache.felix.framework.ServiceRegistrationImpl.unregister(ServiceRegistrationImpl.java:140)
        at org.opencastproject.engage.theodul.manager.impl.EngagePluginManagerImpl.uninstallPlugin(EngagePluginManagerImpl.java:211)
        at org.opencastproject.engage.theodul.manager.impl.EngagePluginManagerImpl.uninstallAll(EngagePluginManagerImpl.java:230)
        at org.opencastproject.engage.theodul.manager.impl.EngagePluginManagerImpl.deactivate(EngagePluginManagerImpl.java:106)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:244)
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41)
        at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:685)
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:529)
        at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:318)
        at org.apache.felix.scr.impl.inject.methods.ActivateMethod.invoke(ActivateMethod.java:308)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.disposeImplementationObject(SingleComponentManager.java:421)
        at org.apache.felix.scr.impl.manager.SingleComponentManager.deleteComponent(SingleComponentManager.java:165)
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.doDeactivate(AbstractComponentManager.java:853)
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.deactivateInternal(AbstractComponentManager.java:825)
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.dispose(AbstractComponentManager.java:589)
        at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.disposeComponents(ConfigurableComponentHolder.java:718)
        at org.apache.felix.scr.impl.BundleComponentActivator.dispose(BundleComponentActivator.java:487)
        at org.apache.felix.scr.impl.Activator.disposeComponents(Activator.java:613)
        at org.apache.felix.scr.impl.Activator.access$300(Activator.java:70)
        at org.apache.felix.scr.impl.Activator$ScrExtension.destroy(Activator.java:451)
        at org.apache.felix.scr.impl.AbstractExtender$1.run(AbstractExtender.java:216)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at org.apache.felix.scr.impl.AbstractExtender.destroyExtension(AbstractExtender.java:238)
        at org.apache.felix.scr.impl.AbstractExtender.bundleChanged(AbstractExtender.java:132)
        at org.apache.felix.scr.impl.Activator.bundleChanged(Activator.java:217)
        at org.apache.felix.framework.EventDispatcher.invokeBundleListenerCallback(EventDispatcher.java:915)
        at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:834)
        at org.apache.felix.framework.EventDispatcher.fireBundleEvent(EventDispatcher.java:516)
        at org.apache.felix.framework.Felix.fireBundleEvent(Felix.java:4579)
        at org.apache.felix.framework.Felix.stopBundle(Felix.java:2626)
        at org.apache.felix.framework.BundleImpl.stop(BundleImpl.java:1038)
        at org.apache.karaf.features.internal.service.BundleInstallSupportImpl.stopBundle(BundleInstallSupportImpl.java:171)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.stopBundle(FeaturesServiceImpl.java:1159)
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:785)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1063)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:998)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

**Server environment:**
 - OS: Any
 - Distribution: Any
- Installation method: RPMs and self build
 - Version: Any before 14.x (13,x happens but this module is off by default)

**Comment by JamesUoM** [comment]
I've enabled the webconsole in our code by adding the follow to `assemblies/dist-develop/pom.xml`

``` 
             <feature>config</feature>
+            <feature>webconsole</feature>
+            <feature>jaas</feature>
             <!-- Opencast external featueres
```

Basically does the same thing but will load the features before theodul, and worth trying to see if it has a different effect. We have removed used theodul modules from our setup.

**Comment by mliradelc** [comment]
Thanks @JamesUoM ! That is a very good workaround!, but still, you need to add it before compilation. Or can you do it in the features file?

**Comment by JamesUoM** [comment]
Yes, You could put in in `org.apache.karaf.features.cfg` mine looks like this:

```
featuresBoot = \
    (wrap), \
    cxf-http-jetty/3.4.6, \
    cxf-commands/3.4.6, \
    log/4.2.15, \
    webconsole/4.2.15, \
    aries-blueprint/4.2.15, \
    feature/4.2.15, \
    eclipselink/2.7.8, \
    jaas/4.2.15, \
    jpa/2.7.3, \
    deployer/4.2.15, \
    bundle/4.2.15, \
    package/4.2.15, \
    scr/4.2.15, \
    cxf-jaxrs/3.4.6, \
    opencast-allinone/12.0.0.SNAPSHOT, \
    framework/4.2.15, \
    system/4.2.15, \
    shell/4.2.15, \
    service/4.2.15, \
    jndi/4.2.15, \
    http/4.2.15, \
    http-whiteboard/4.2.15, \
    config/4.2.15, \
    transaction/2.0.0
```

Let me know if it works.

**Comment by gregorydlogan** [comment]
Adding `jaas` prior to the OC feature does not work for me.  It'll boot, but later installing `webconsole` causes the same issue.  Putting both into the features works as expected.

I'm going to guess this is JAAS + Spring related.  When you look at the logs after `jaas` has 'installed' (note: it doesn't successfully hot-install), you see [this line](https://github.com/opencast/opencast/blob/r/12.x/modules/engage-theodul-core/src/main/java/org/opencastproject/engage/theodul/manager/impl/EngagePluginManagerImpl.java#L92) repeating, suggesting that the `RestPublisher`'s bundle context is somehow null.  If I had to guess, when `jaas` gets installed and a bunch of things get shut down the `RestPublisher` should be one of them, and it's either not being shut down at all, or it's failing/hanging part way through.  Looking at the logs, I don't see it shut down, nor do I see it boot back up...

**Comment by mliradelc** [comment]
Like Greg, I had the same issue, but backwards.

@JamesUoM Works like a Charm. I will write a PR on how to enable the webconsole by code or using the Karaf features file. The hot installation make a lot of problems with opencast.

**Comment by gregorydlogan** [comment]
IIRC, the major reason the webconsole was disabled by default was that it let anyone with any kind of admin role (ROLE_COURSE_ADMIN, etc) have access to it.  I don't remember if we tested to see if we could restrict that, but when you're prepping your PR make sure that console is restricted to ROLE_ADMIN by default.

**Comment by mtneug** [comment]
> IIRC, the major reason the webconsole was disabled by default was that it let anyone with any kind of admin role (ROLE_COURSE_ADMIN, etc) have access to it. I don't remember if we tested to see if we could restrict that, but when you're prepping your PR make sure that console is restricted to ROLE_ADMIN by default.

Wait the suggestion is not to activate this by default right? I understand @mliradelc comment to be only documentation.

**Comment by gregorydlogan** [comment]
I misread Max's last. +1 for docs. -1 for enable by default, at least until we're sure it's only ROLE_ADMIN is the only one with access. And even then I'm not sure it's a good idea.

**Comment by mliradelc** [comment]
Indeed, I'm thinking in only write a “How to” guide to add it your own OC build or how to include it later in karaf features file. 😉

**Comment by mliradelc** [comment]
@JamesUoM Now I tried in our cluster, but I've found that only `jaas` is installed but not the `webconsole` when I add them to the `org.apache.karaf.features.cfg`

Here is the installation order, also I tried setting `webconsole` at the end but no success.

```
featuresBoot = \
    (wrap), \
    cxf-http-jetty/3.4.6, \
    package/4.2.16, \
    scr/4.2.16, \
    cxf-commands/3.4.6, \
    log/4.2.16, \
    jaas/4.2.15, \
    cxf-jaxrs/3.4.6, \
    aries-blueprint/4.2.16, \
    framework/4.2.16, \
    system/4.2.16, \
    feature/4.2.16, \
    shell/4.2.16, \
    eclipselink/2.7.8, \
    service/4.2.16, \
    opencast-pluginmanager/13.0.0.SNAPSHOT, \
    jndi/4.2.16, \
    http/4.2.16, \
    jpa/2.7.3, \
    deployer/4.2.16, \
    http-whiteboard/4.2.16, \
    opencast-admin/13.0.0.SNAPSHOT, \
    opencast-security-jwt/13.0.0.SNAPSHOT, \
    bundle/4.2.16, \
    config/4.2.16, \
    transaction/2.0.0, \
    webconsole/4.2.15
```

Can I enable the log for `Karaf features` when try to set each feature?

**Comment by JamesUoM** [comment]
I made it permanent by adding them to the dev assemble pom file `assemblies/dist-develop/pom.xml`

```
             <feature>service</feature>
             <feature>system</feature>
             <feature>config</feature>
+            <feature>webconsole</feature>
+            <feature>jaas</feature>
             <!-- Opencast external features -->
             <feature>jpa</feature>
             <feature>http</feature>
```

**Comment by mliradelc** [comment]
Close this issue. Seems to be fix with #5151
